components/setup/DockerInstructions.tsx

```diff
@@ -0,0 +1,207 @@
+	return filteredRepos;
+}
+
+function formatRepoListInSaveSetupArgsForm(repos: RepoIdentifier[], install_id: string) {
+	// Group by repo_owner and generate setup args
+	const setupArgsMap: Map<string, SetupReposApiBodyArgs> = new Map();
+	repos.forEach(repo => {
+		const key = repo.repo_owner;
+		if (!setupArgsMap.has(key)) {
+			setupArgsMap.set(key, {
+				owner: repo.repo_owner,
+				provider: repo.repo_provider,
+				repos: [],
+				installationId: install_id
+			});
+		}
+		setupArgsMap.get(key)!.repos.push(repo.repo_name);
+	});
+
+	return Array.from(setupArgsMap.values());
+}
+const DockerInstructions: React.FC<DockerInstructionsProps> = ({ userId, selectedInstallationType, selectedProvider, session }) => {
+	const [isCopied, setIsCopied] = useState<boolean>(false);
+	const [isButtonDisabled, setIsButtonDisabled] = useState<boolean>(false);
+	const [selfHostingCode, setSelfHostingCode] = useState<string>("Generating topic name, please try refreshing if you keep seeing this...");
+	const [selectedRepos, setSelectedRepos] = useState<RepoIdentifier[]>([]);
+	const [allRepos, setAllRepos] = useState<RepoIdentifier[]>([]);
+	const [isRepoSelectionDone, setIsRepoSelectionDone] = useState<boolean>(false);
+	const [installId, setInstallId] = useState<string | null>(null);
+
+	useEffect(() => {
+		axios.post('/api/dpu/pubsub', { userId }).then(async (response) => {
+			if (response.data.installId) {
+				setInstallId(response.data.installId);
+				if (selectedInstallationType === 'individual' && selectedProvider === 'github') {
+					if (!session) {
+						console.error(`[DockerInstructions] could not get session for userId: ${userId}`);
+						return;
+					}
+					if (isRepoSelectionDone) {
+						setSelfHostingCode(`
+docker pull asia.gcr.io/vibi-prod/dpu/dpu &&\n
+docker run -e INSTALL_ID=${response.data.installId} \\
+-e PROVIDER=<your_provider_here> \\
+-e GITHUB_PAT=<Your gh cli token> \\
+asia.gcr.io/vibi-prod/dpu/dpu
+					`);
+					}
+					else {
+						let providerReposForUser = await getUserReposForProvider(session, 'github');
+						setAllRepos(providerReposForUser);
+					}
+				} else if (selectedInstallationType === 'individual' && selectedProvider === 'bitbucket') {
+					setSelfHostingCode(`
+Coming Soon!
+					`)
+				} else {
+					setSelfHostingCode(`
+docker pull asia.gcr.io/vibi-prod/dpu/dpu &&\n
+docker run -e INSTALL_ID=${response.data.installId} asia.gcr.io/vibi-prod/dpu/dpu
+					`);
+				}
+			}
+		}).catch((error) => {
+			setSelfHostingCode(`Unable to get topic name for user\nPlease refresh this page and try again.`);
+			console.error(`[DockerInstructions] Unable to get topic name for user ${userId} - ${error.message}`);
+		});
+	}, [userId, selectedInstallationType, selectedProvider]);
+
+	const handleCopyClick = () => {
+		setIsButtonDisabled(true);
+	};
+
+	const handleCopy = () => {
+		setIsCopied(true);
+		setIsButtonDisabled(false);
+	};
```

Comment:
Should be in the `CodeWithCopyButton` component. They need not be created again for different callers unnecessarily.